File: 算法基础课(修订版)/笔试拿分技巧/P14160/verify_acm.java
import java.util.*;

class Solution {
    static final long MOD = 1_000_000_007;

    static class BIT {
        int size;
        long[] tree;

        BIT(int n) {
            size = n;
            tree = new long[n + 1];
        }

        void add(int x, long value) {
            while (x <= size) {
                tree[x] = (tree[x] + value) % MOD;
                x += x & -x;
            }
        }

        long query(int x) {
            long result = 0;
            while (x > 0) {
                result = (result + tree[x]) % MOD;
                x -= x & -x;
            }
            return result;
        }
    }

    public long solve(ArrayList<Long> a) {
        ArrayList<Long> alls = new ArrayList<>(new HashSet<>(a));
        Collections.sort(alls, Collections.reverseOrder());

        BIT bit = new BIT(alls.size());
        long result = 0;
        for (long value : a) {
            int id = Collections.binarySearch(alls, value, Collections.reverseOrder());
            if (id < 0) {
                continue;
            }
            id++;
            long sum = bit.query(id - 1);
            long res = (sum + 1) % MOD;
            result = (result + res) % MOD;
            bit.add(id, res);
        }
        return result;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        ArrayList<Long> a = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            a.add(sc.nextLong());
        }

        Solution solution = new Solution();
        System.out.println(solution.solve(a));
        sc.close();
    }
}

File: 算法基础课(修订版)/笔试拿分技巧/P14160/verify_acm.cc
#include <bits/stdc++.h>
using namespace std;

const int MOD = 1e9 + 7;

struct BIT {
    int size;
    vector<long long> tree;

    BIT(int n) : size(n), tree(n + 1, 0) {}

    void add(int x, long long value) {
        while (x <= size) {
            tree[x] = (tree[x] + value) % MOD;
            x += x & -x;
        }
    }

    long long query(int x) const {
        long long res = 0;
        while (x > 0) {
            res = (res + tree[x]) % MOD;
            x -= x & -x;
        }
        return res;
    }
};

class Solution {
public:
    long long solve(vector<int>& a) {
        vector<int> alls = a;
        sort(alls.begin(), alls.end(), greater<int>());
        alls.erase(unique(alls.begin(), alls.end()), alls.end());

        BIT bit((int)alls.size());
        long long ans = 0;
        for (int x : a) {
            int id = lower_bound(alls.begin(), alls.end(), x, greater<int>()) - alls.begin() + 1;
            long long sum = bit.query(id - 1);
            long long res = (sum + 1) % MOD;
            ans = (ans + res) % MOD;
            bit.add(id, res);
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> a(n);
    for (int i = 0; i < n; ++i) {
        cin >> a[i];
    }

    Solution solution;
    cout << solution.solve(a);
    return 0;
}
